design/README.md

```diff
@@ -0,0 +1,285 @@
+
+### Managed Homestar FaaS
+
+Terri wants to register a function on Odd Functions. They go to the Fission CLI, and signs up for the service with their email, and registers the username `terri2023`. They're prompted to set up a Passkey (which will be their root DID). The Fission FaaS service then delegates 1000 credits to `did:key:zTerriPasskey`.
+
+# Approach
+
+## Event Sourcing
+
+Instead of recording the `createdAt` and latest `modifiedAt`, the Fission server records immutable events. This stores more data in the database, but has multiple advantages:
+
+- Schema migrations can be reconstituted performed directly or by playing over the log
+- Post hoc analysis and debugging is possible
+- Nondestructive updates makes fixing bugs in-place significantly easier
+
+## Resource DIDs
+
+In the previous (Haskell) server, the user's DID was registered directly in the (central) database. This approach has a few drawbacks:
+
+- Requires a DB lookup to find the associated owner for an account (or other resource)
+- Doesn't provide a solution to uniquely identifying a single resource in a one-to-many relationship
+- Has ACL semantics at the final step of the delegation (which is not unheard of in capability modeling)
+
+Instead, if every resource (include the account itself) is uniquely identified by a DID, then UCANs can be used throughout without resorting to an external ownership list. With this approach, the keys for the resource may be stored in a KMS, or ideally _dropped completely_ after delegating root access to one or more controllers. To solve for recovery and a server-controlled revocation fail-safe, a key may be kept offline (i.e. a "cold DID") and delegated to.
+
+Here is one example of such a DID hierarchy:
+
+``` mermaid
+flowchart
+    RootServerDID --> 2023ServerDID
+    2023ServerDID --> Jan2023ServerDID
+    2023ServerDID --> Feb2023ServerDID
+    2023ServerDID --> Mar2023ServerDID
+
+    RootServerDID --> 2024ServerDID
+    2024ServerDID --> Jan2024ServerDID
+    2024ServerDID --> Feb2024ServerDID
+    2024ServerDID --> Mar2024ServerDID
+
+    RootServerDID --> ...
+```
+
+Here is one such flow, with a cold server DID and monthly hot server DID:
+
+``` mermaid
+sequenceDiagram
+    participant AccountDID
+    participant ServerRootDID
+    participant JanServerDID
+    actor UserRootDID
+
+    Note over AccountDID, UserRootDID: Earlier Server Setup
+    ServerRootDID ->> JanServerDID: UcanDelegate(any)
+
+    Note over AccountDID, UserRootDID: Account Creation
+    UserRootDID ->>+ JanServerDID: newAccount(RootDID)
+        JanServerDID ->>+ AccountDID: spawn!(owners: [ServerRootDID])
+        AccountDID ->> ServerRootDID: UcanDelegate (parenthood)
+        AccountDID -->> AccountDID: dropPrivateKey!
+        ServerRootDID -->> JanServerDID: (recall that this UCAN delegation exists)
+        JanServerDID ->> UserRootDID: UcanDelegate
```

Comment:
I don't 100% understand how this encodes in the form of UCANs yet.

The way I understand this, this forms a closed loop:
```mermaid
flowchart TD
    A[ServerRootDID] -->|"UcanDelegate(any)"| JanServerDID
    JanServerDID -->|"spawn!(...)"| AccountDID
    AccountDID -->|"UcanDelegate(parenthood)"| B[ServerRootDID]
```

The loop starts and finally ends at `ServerRootDID`.

Then there's this other UCAN delegation chain:
```mermaid
flowchart TD
    A[ServerRootDID] -->|"UcanDelegate(any)"| JanServerDID
    JanServerDID -->|UcanDelegate| UserRootDID
    UserRootDID -->|UcanDelegate| B[ServerRootDID]
```

Also starting and ending in `ServerRootDID`.

Now, in theory these loops could cross over to one another via the `AccountDID -->|UcanDelegate(parenthood)| ServerRootDID`, but I don't know how to encode that as a UCAN.

For principal alignment, it'd need to be in the `prf` of the `ServerRootDID -> JanServerDID` from way above. But obviously that's impossible, since that UCAN is created at an earlier time.

Put differently:
From the viewpoint of `UserRootDID`, in the diagram it can only "attach" arrows (create valid UCANs) that start from the `UserRootDID` node.
Since it's not part of the loop involving `AccountDID` at all, it can't make use of anything in there.

From the viewpoint of `JanServerDID`, it is indeed part of the diagram involving `AccountDID`. However, it's not "downstream" of anything useful to it. The only thing that's issued to it is `UcanDelegate(any)` from the `ServerRootDID`. It can't put the `UcanDelegate(parenthood)` into any `prf` of any UCAN it creates, since the audience for that UCAN is `ServerRootDID`.

---

I really hope this makes sense. It's hard to explain!

Reply:
> The way I understand this, this forms a closed loop:

Correct! That last step is an Invocation, but we don't have invocations today. It probably just should say "invocation", though — will fix.

> the parenthood capability

Just to make sure that we're talking about the same thing, the is no "parenthood capability". It's "by parenthood" AKA "I created you so immedietly delegate to me". I see how the way it's currently written can be confusing; I'll update that text.

> cross over

I don't follow. What needs to cross?

Reply:
Updated to include invocation

<img width="690" alt="Screenshot 2023-09-26 at 12 16 27" src="https://github.com/fission-codes/fission-server/assets/1052016/41cfb176-2551-4c1d-844e-65f56c7ef1ae">

Reply:
I was trying to keep the diagram minimal, but here's revocation explicitly too

<img width="704" alt="Screenshot 2023-09-26 at 12 41 08" src="https://github.com/fission-codes/fission-server/assets/1052016/7fa25e02-fb8e-42a3-b957-8f7f7a36798e">